### PR TITLE
ARROW-17334: [Java] Add maven-source-plugin to package source code

### PR DESCRIPTION
**Pull request by jinchengchenghh**

### CODE REVIEW

**Comment by pitrou** [comment]
@davisusanibar @lwhite1 Would one of you want to review this?

File: java/pom.xml [hunk 1]
```diff
@@ -358,6 +358,19 @@
...
+        <executions>
+          <execution>
+            <id>attach-sources</id>
+            <phase>package</phase>
+            <goals>
+              <goal>jar-no-fork</goal>
+            </goals>
+          </execution>
+        </executions>
+      </plugin>
```
Comment (davisusanibar) [inline java/pom.xml:373]: What about [maven-dependency-plugin](https://maven.apache.org/plugins/maven-dependency-plugin/) it tells Maven to resolve all dependencies and their source attachments with `mvn dependency:sources --debug`?

```
[INFO] --- maven-dependency-plugin:2.8:sources (default-cli) @ wks-reproduce ---
Downloaded from central: https://repo.maven.apache.org/maven2/org/apache/arrow/arrow-vector/9.0.0/arrow-vector-9.0.0-sources.jar (946 kB at 5.1 MB/s)
```

Comment (jinchengchenghh) [reply java/pom.xml:373]: For release version 9.0.0, it works well. How can I  get the source for the current branch master, I compiled 10.0.0-SNAPSHOT by myself.

**Comment by jinchengchenghh** [comment]
Good! Much thanks! After follow this example, [example](https://github.com/apache/arrow/blob/acf25f764d6076b46f18bbbe2b04b0a5d970d448/ci/scripts/java_full_build.sh#L50:L60), I can get the source jar.

**Pull request closed without merging**